### PR TITLE
Eng 81: List Surveys. Edit Surveys. Create theme, update survey status

### PR DESCRIPTION
**Pull request by kaf-lamed-beyt**
## stuffs done
- [x] List all surveys in a project
- [x] edit survey (name, themes, survey settings/preferences)
- [x] delete theme
- [x] delete a survey and change survey status.
- [x] update createSurvey so new surveys are stored in cache

@AfeezGL, when you're free. Please review.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -23,9 +23,9 @@
     "[javascript][typescript][javascriptreact][typescriptreact]": {
         "editor.formatOnSave": true,
         "editor.codeActionsOnSave": {
-            "source.organizeImports": "always",
-            "source.formatDocument": "always",
-            "source.fixAll.eslint": "always"
+            "source.organizeImports": "explicit",
+            "source.formatDocument": "explicit",
+            "source.fixAll.eslint": "explicit"
```
Comment (jdamilola) [inline .vscode/settings.json:28]: @kaf-lamed-beyt Why did you change this?

File: apps/frontend/src/modules/projects/hooks/useTheme.ts [hunk 2]
```diff
@@ -19,16 +20,17 @@ export type ColorScheme = {
 };
 
 export const useThemes = () => {
-    const { user } = useUserState();
     const { workspace } = useWorkspace();
     const [createThemeMutation] = useProjectThemeCreateMutation();
     const [updateThemeMutation] = useProjectThemeUpdateMutation();
     const { updateSurvey } = useSurvey();
```
Comment (jdamilola) [inline apps/frontend/src/modules/projects/hooks/useTheme.ts:26]: This will eventually create a cyclic dependency. It will make this hook hard to test later. Why do we need `useSurvey` hook here anyway? We should probably make two calls whenever we create/delete a theme.

File: apps/frontend/src/modules/projects/hooks/useTheme.ts [hunk 3]
```diff
@@ -19,16 +20,17 @@ export type ColorScheme = {
 };
 
 export const useThemes = () => {
```
Comment (jdamilola) [inline apps/frontend/src/modules/projects/hooks/useTheme.ts:22]: @kaf-lamed-beyt This should be `useTheme` since we are dealing with just a single theme

File: apps/frontend/src/modules/surveys/components/studio/create/editor-panel/DesignEditor/components/PresetThemes.tsx [hunk 4]
```diff
@@ -1,4 +1,5 @@
 import { useThemes } from "@/modules/projects/hooks/useTheme";
```
Comment (jdamilola) [inline apps/frontend/src/modules/surveys/components/studio/create/editor-panel/DesignEditor/components/PresetThemes.tsx:1]: @kaf-lamed-beyt We should probably rethink the implementation of theme. Tight-coupling it with survey doesn't seem like the right way to go. Remember, it's supposed to be under project which means it will still exist even if Survey goes away or if we have other in-app content like `Poll`, `Quick`, `Push` e.t.c. We should probably move it to project module, then utilise event handlers to update things in other modules. Let me know if you need more explanations.

File: .vscode/settings.json (hunk 1 above)
Comment (kaf-lamed-beyt) [reply .vscode/settings.json:28]: I made a mistake while setting up vscode on my PC. so I went and copied this vscode settings.json file as is from my branch.

File: apps/frontend/src/modules/projects/hooks/useTheme.ts (hunk 3 above)
Comment (kaf-lamed-beyt) [reply apps/frontend/src/modules/projects/hooks/useTheme.ts:22]: Aha! I see. I'll change it now.

File: apps/frontend/src/modules/surveys/components/studio/create/editor-panel/DesignEditor/components/PresetThemes.tsx (hunk 4 above)
Comment (kaf-lamed-beyt) [reply apps/frontend/src/modules/surveys/components/studio/create/editor-panel/DesignEditor/components/PresetThemes.tsx:1]: Makes sense! I left a message for you BTW.

File: apps/frontend/src/modules/projects/hooks/useTheme.ts (hunk 2 above)
Comment (jdamilola) [reply apps/frontend/src/modules/projects/hooks/useTheme.ts:26]: @kaf-lamed-beyt I think we should move the `updateSurvey` to the function the theme is created. The reason is that we should be able to create a theme without a survey. A quick example, in project settings, we'll have a manage themes feature. Updating survey in this hook will break this feature.

File: apps/frontend/src/modules/projects/components/ThemeCard.tsx [hunk 5]
```diff
@@ -26,6 +27,11 @@ export const ThemeCard = ({
         return theme.colorScheme[key];
     };
 
+    const editTheme = () => {
+        useStudioStore.setState({ editTheme: true });
```
Comment (jdamilola) [inline apps/frontend/src/modules/projects/components/ThemeCard.tsx:31]: @kaf-lamed-beyt This is not right. We shouldn't be accessing studio in project module. It breaks dependency flow, and will make it difficult to test this component. Studio should be able to reference Theme but not vice-versa.

File: apps/frontend/src/modules/surveys/states/studio.ts [hunk 6]
```diff
@@ -2,6 +2,7 @@ import { create } from "zustand";
 
 export type StudioState = {
     studioModeIsActive: boolean;
+    editTheme: boolean;
```
Comment (jdamilola) [inline apps/frontend/src/modules/surveys/states/studio.ts:5]: @kaf-lamed-beyt Why do we need this state?

File: apps/frontend/src/modules/projects/components/ThemeCard.tsx (hunk 5 above)
Comment (kaf-lamed-beyt) [reply apps/frontend/src/modules/projects/components/ThemeCard.tsx:31]: Since the state is needed in the studio that was why I placed it in the state local to the studio. Should I move it into the project module? Perhaps creating a `states` dir in it?

File: apps/frontend/src/modules/surveys/states/studio.ts (hunk 6 above)
Comment (kaf-lamed-beyt) [reply apps/frontend/src/modules/surveys/states/studio.ts:5]: @jdamilola, It is referenced here. 

https://github.com/IntegraflowHQ/integraflow/blob/df28ca66d3342d46170895f4a03b4b0203327e4d/apps/frontend/src/modules/surveys/components/studio/create/editor-panel/DesignEditor/index.tsx#L182-L201

I'm using a global state with Zustand, so we don't end up having race conditions from drilling props.

Comment (kaf-lamed-beyt) [reply apps/frontend/src/modules/surveys/states/studio.ts:5]: The idea behind this is so, the UI of the color picker &mdash; the buttons, specifically &mdash; matches the action a user performs. 

If they try to click on the Pen icon, to edit a theme, both buttons &mdash; Update theme and Delete Theme &mdash; should be rendered. If not, one button should be rendered when they click on the "New Theme" button, and the text should be "Create Theme" since they'll be creating a new theme. You can't delete a theme that hasn't been created.

Comment (kaf-lamed-beyt) [reply apps/frontend/src/modules/surveys/states/studio.ts:5]: I'm still working on the feedback you gave me yesterday.  By evening or Night, I should've updated this PR with the fix. If I don't find a headway... I'll let you know.

**Comment by kaf-lamed-beyt** [comment]
@jdamilola, I added `<SurveyProvider />` a couple of seconds ago. Let me know how it looks, thanks!

File: apps/frontend/src/main.tsx [hunk 7]
```diff
@@ -63,22 +64,43 @@ const router = createBrowserRouter([
                     },
                     {
                         path: "projects/:projectSlug/surveys",
-                        element: <Surveys />,
+                        element: (
```
Comment (jdamilola) [inline apps/frontend/src/main.tsx:67]: @kaf-lamed-beyt This is incorrect. All survey related components should be under a single `SurveyProvider`. Did you try the code I shared with you earlier?

File: apps/frontend/src/modules/surveys/SurveyProvider.tsx [hunk 8]
```diff
@@ -73,20 +86,61 @@ export const SurveyProvider = ({ children }: SurveyProviderProp) => {
...
+                status: edge?.node?.status,
+                createdAt: edge?.node?.createdAt,
+                name: edge?.node?.name ? edge?.node?.name : "Untitled survey",
+                creator: {
+                    email: edge?.node?.creator.email,
+                    fullName: `${edge?.node?.creator.firstName} ${edge?.node?.creator.lastName}`,
+                },
+            };
+        })
+        .sort((a, b) => {
```
Comment (jdamilola) [inline apps/frontend/src/modules/surveys/SurveyProvider.tsx:137]: @kaf-lamed-beyt Why are you sorting locally? All form of sorting should be from the backend.

File: apps/frontend/src/modules/surveys/SurveyProvider.tsx [hunk 9]
```diff
@@ -73,20 +86,61 @@ export const SurveyProvider = ({ children }: SurveyProviderProp) => {
     const [deleteSurveyMutation, { loading: pendingDeletion }] =
         useSurveyDeleteMutation();
 
+    const [getSurvey, { loading: loadingSurvey, data: survey }] =
+        useGetSurveyLazyQuery({
+            variables: {
+                slug: surveySlug,
```
Comment (jdamilola) [inline apps/frontend/src/modules/surveys/SurveyProvider.tsx:92]: @kaf-lamed-beyt This variables should be added when you trigger `getSurvey` in the `useEffect`.

File: apps/frontend/src/main.tsx (hunk 7 above)
Comment (kaf-lamed-beyt) [reply apps/frontend/src/main.tsx:67]: Yes, I did. 

SurveyProvider expects a `children` prop. Appending it like so: `<SurveyProvider />` throws an error. Can't remember the exact error in the console. But, the page became blank, nothing rendered.

File: apps/frontend/src/modules/surveys/SurveyProvider.tsx (hunk 8 above)
Comment (kaf-lamed-beyt) [reply apps/frontend/src/modules/surveys/SurveyProvider.tsx:137]: Yes. I added this as a placeholder until I'm able to do so.

File: apps/frontend/src/modules/surveys/SurveyProvider.tsx [hunk 10]
```diff
@@ -73,20 +86,61 @@ export const SurveyProvider = ({ children }: SurveyProviderProp) => {
...
-            slug: surveySlug,
+            first: SURVEYS_PER_PAGE,
+            sortBy: {
+                field: SurveySortField.Name,
+                direction: OrderDirection.Desc,
+            },
+        },
+        context: {
+            headers: {
+                Project: workspace?.project.id,
```
[the remark this answers is not in the document]
Comment (kaf-lamed-beyt) [reply apps/frontend/src/modules/surveys/SurveyProvider.tsx:111]: Oh! I thought we needed it for the context or so. Noted. I'll have it removed.

File: apps/frontend/src/main.tsx (hunk 7 above)
Comment (jdamilola) [reply apps/frontend/src/main.tsx:67]: @kaf-lamed-beyt Because you need an `Outlet` from `react-router-dom`. That should be its child. So it should be like;

```export default function SurveyContainer() {
    return (
        <SurveyProvider>
                <Outlet />
        </SurveyProvider>
    );
}```

Then use `SurveyContainer` in `main.tsx`.

File: apps/frontend/src/modules/surveys/SurveyProvider.tsx (hunk 9 above)
Comment (kaf-lamed-beyt) [reply apps/frontend/src/modules/surveys/SurveyProvider.tsx:92]: @jdamilola 

Like the `getSurvey` function should accept args? Or better still, can you show me an example, please?

File: apps/frontend/src/main.tsx (hunk 7 above)
Comment (kaf-lamed-beyt) [reply apps/frontend/src/main.tsx:67]: Aha! I see. Thank you.

**Comment by kaf-lamed-beyt** [comment]
@jdamilola, rendering the selected survey theme works now.

This became possible when I extended its value from the provider. Instead of trying to access it in the component. The same thing for the survey settings.

I think it sits well with the pattern &mdash; of making a module self-sufficient &mdash; you discussed with us.

File: apps/frontend/src/main.tsx [hunk 11]
```diff
@@ -59,40 +60,26 @@ const router = createBrowserRouter([
...
-                        element: (
-                            <SurveyProvider>
-                                <Surveys />
-                            </SurveyProvider>
-                        ),
+                        path: "projects/:projectSlug",
+                        element: <SurveyShell />,
+                        children: [
+                            {
+                                path: "get-started",
```
Comment (jdamilola) [inline apps/frontend/src/main.tsx:67]: @kaf-lamed-beyt The Onboarding feature should be outside the SurveyShell.

File: backend/integraflow/graphql/survey/resolvers.py [hunk 12]
```diff
@@ -46,6 +46,8 @@ def resolve_survey(info, id=None, slug=None):
 
     lookup = None
 
+    print("SLUG", id, slug, project.pk)
```
Comment (AfeezGL) [inline backend/integraflow/graphql/survey/resolvers.py:49]:
```suggestion
```

File: apps/frontend/src/modules/surveys/SurveyProvider.tsx [hunk 13]
```diff
@@ -0,0 +1,609 @@
...
+                },
+            },
+            notifyOnNetworkStatusChange: true,
+            fetchPolicy: "network-only",
+        });
+    }, [getSurveyList, projectSlug, workspace?.project?.id]);
+
+    const createSurvey = React.useCallback(
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
+        async (_template?: string) => {
```
Comment (AfeezGL) [inline apps/frontend/src/modules/surveys/SurveyProvider.tsx:165]: @kaf-lamed-beyt Is the `_template` parameter being used anywhere?

File: apps/frontend/src/modules/surveys/SurveyProvider.tsx [hunk 14]
```diff
@@ -0,0 +1,609 @@
...
+
+                update: (cache, { data }) => {
+                    if (!data?.surveyCreate?.survey) return;
+
+                    cache.modify({
+                        fields: {
+                            surveys(existingSurveys = []) {
+                                const newSurveyRef = cache.writeFragment({
+                                    data: data.surveyCreate?.survey,
+                                    fragment: SURVEY_QUESTION,
```
Comment (AfeezGL) [inline apps/frontend/src/modules/surveys/SurveyProvider.tsx:229]: @kaf-lamed-beyt Why is the fragment here `SURVEY_QUESTION`?

File: apps/frontend/src/ui/Dialog/Dialog.tsx [hunk 15]
```diff
@@ -31,7 +31,7 @@ export const DialogContent = forwardRef<HTMLDivElement, Props>(
                                 {title}
                             </DialogPrimitive.Title>
                             <DialogPrimitive.Close aria-label="Close">
-                                <XIcon />
+                                <XIcon className="absolute right-[18px] top-[14px] inline-flex" />
```
Comment (AfeezGL) [inline apps/frontend/src/ui/Dialog/Dialog.tsx:34]: @kaf-lamed-beyt I can see you moved this icon up, I thought it was in line with the design, why was it moved?

File: apps/frontend/src/ui/Dialog/Dialog.tsx [hunk 16]
```diff
@@ -43,7 +43,7 @@ export const DialogContent = forwardRef<HTMLDivElement, Props>(
                     <>
                         <div className="flex justify-end">
                             <DialogPrimitive.Close aria-label="Close">
-                                <XIcon />
+                                <XIcon className="absolute right-[18px] top-[14px] inline-flex" />
```
Comment (AfeezGL) [inline apps/frontend/src/ui/Dialog/Dialog.tsx:46]: @kaf-lamed-beyt and this too.

File: apps/frontend/src/modules/projects/components/ThemesMenu.tsx
Comment (AfeezGL) [inline apps/frontend/src/modules/projects/components/ThemesMenu.tsx:1]: @kaf-lamed-beyt are we still using the `ThemesMenu` component?

File: apps/frontend/src/modules/surveys/SurveyProvider.tsx [hunk 17]
```diff
@@ -0,0 +1,609 @@
...
+export type SurveyContextValues = {
+    surveysOnPage: number;
+    totalSurveys: number | null | undefined;
+    pageInfo: PageInfo | undefined;
+    loading: boolean;
+    error: ApolloError | undefined;
+    openQuestion: string;
+    surveyName: string | undefined;
+    selectedSurveyTheme: string | undefined;
+    surveyId: string | undefined;
```
Comment (jdamilola) [inline apps/frontend/src/modules/surveys/SurveyProvider.tsx:52]: @kaf-lamed-beyt Why do we have all these variables? I thought they are all already in `survey`. Can you remove `surveyId`, `surveyName`, `selectedSurveyTheme`, `pageInfo` e.t.c. They are alternative ways to access these values.

**Comment by kaf-lamed-beyt** [comment]
Hi @jdamilola,  I've included the changes to `SurveyProvider` as requested. Kindly take a look when you're free.

@AfeezGL, I made a tiny change to the `<GlobalSpinner />`. One that makes it a bit flexible. I added a prop to conditionally render the spinner on the screen instead of removing the whole UI, and replacing it with the spinner. 

| Before | After |
| --------- | ------ |
| global | in-studio |
